src/ixdat/tools.py

```diff
@@ -71,3 +80,159 @@ def list_is_close(list_one, list_two):
+            that this deprecation will remain soft for the foreseeable future.
+        remove_release (str): The release with which the deprecated functionality is
+            planned for deletion
+        kwarg_name (str): If given, is the name of the keyword argument which is
+            deprecated. If not given it is assumed that the entire callable is
+            deprecated.
+
+    Examples:
+
+    Used to deprecate a class::
+     @deprecate("1.2.3", "Please use `MyNewClass` instead", "1.4.0", "2.0.0")
+     class MyClass:
+         ...
+
+    Used to deprecate a method or an argument in a method::
+        class MyClass:
+
+            @deprecate("1.2.3", "Please use `mynewmethod` instead", "1.4.0", "2.0.0")
+            def mymethod(self):
+                ...
+
+            @classmethod
+            @deprecate(
+                "1.2.3",
+                "Please use `kwargb` instead",
+                "1.4.0",
+                "2.0.0",
+                kwarg_name="kwarga"
+            )
+            def mymethod(cls, arg, kwarga=None, kwargb=None):
+                ...
+
+    .. note::
+       In the example above, when this decorator is applied to a class method or
+       static method, it must be applied as the first decorator (closest to the def).
+
+    """
+
+    def decorator(callable_):
+        """Decorate a callable with"""
+        # Form an identity string for the object which is being decorated, which is used
+        # in the message to the user
+        if inspect.isclass(callable_):
+            identity = f"class '{callable_.__qualname__}'"
+        else:
+            if "." in callable_.__qualname__:
+                identity = f"method '{callable_.__qualname__}'"
+            else:
+                identity = f"function '{callable_.__qualname__}'"
+
+        compound_message = _construct_deprecation_message(
+            identity,
+            last_supported_release,
+            update_message,
+            hard_deprecation_release,
+            remove_release,
+            kwarg_name,
+        )
+
+        # Get the argument signature of the callable
+        callable_signature = inspect.signature(callable_)
+
+        @wraps(callable_)
+        def inner_function(*args, **kwargs):
+            # Form bound arguments, so both args and kwargs gets mapped to their name
+            bound_arguments = {}
+            if kwarg_name:
+                bound_arguments = callable_signature.bind(*args, **kwargs).arguments
+
+            # If something deprecated is being used, issue the warning or exception
+            if kwarg_name is None or kwarg_name in bound_arguments:
+                if hard_deprecation_release is not None and version.parse(
+                    __version__
+                ) >= version.parse(hard_deprecation_release):
+                    raise DeprecationError(compound_message)
+                else:
+                    warnings.warn(compound_message, DeprecationWarning, stacklevel=2)
```

Comment:
what is `stacklevel`? A comment here could be useful.

Reply:
Ah, agreed! The stack level refers to the call stack. This is a way to make sure that the reference in the warning text about where the warning was issued, will refer to the correct place in the decorated function and not to the place here in the inner function, which would be useless. I will add a comment.